### ISSUE TITLE
Hierarchical Templates for Array Encoding using $* for [] on URI

### ISSUE THREAD

**Issue by vitorpamplona** [Consensas/jsonxt#6]
Here’s an idea to encode arrays:

Given a JSON:
```json
{
    "people": [{
        "type": "Person",
        "given": "Jane",
        "family": "Doe",
        "ids": [{
            "type": "IdentificationDocument",
            "name": "MDL",
            "number": "23EFG3D5CC"
        }, {
            "type": "IdentificationDocument",
            "name": "Passport",
            "number": "YNW32ND"
        }]
    }, {
        "type": "Person",
        "given": "John",
        "family": "Doe", 
        "ids": [{
            "type": "IdentificationDocument",
            "name": "MDL",
            "number": "94568DDS1"
        }, {
            "type": "IdentificationDocument",
            "name": "Passport",
            "number": "2SC6223"
        }]
    }]
}
```
Templates.json
```json
{
  "data:1": {
    "columns": [
       {"path": "people", "encoder": "array", "encoder_param": "person:1"},
    ],
    "template": {}
  },
  "person:1": {
    "columns": [
       {"path": "given", "encoder": "string"},
       {"path": "family", "encoder": "string"},
       {"path": "ids", "encoder": "array", "encoder_param": "identification:1"},
    ],
    "template": {
        "type": "Person"
    }
  },
  "identification:1": {
    "columns": [
       {"path": "name", "encoder": "string"},
       {"path": "number", "encoder": "string"}
    ],
    "template": {
        "type": "IdentificationDocument"
    }
  }
}
```

Produces:

```JXT::data:1:$Jane/Doe/$MDL/23EFG3D5CC/Passport/YNW32ND*/John/Doe/$MDL/94568DDS1/Passport/2SC6223**```

Breaking down the URI: 

```
JXT::data:1:
  $
    Jane/
    Doe/
    $
      MDL/23EFG3D5CC/
      Passport/YNW32ND
    */
    John/
    Doe/
    $
      MDL/94568DDS1/
      Passport/2SC6223
    *
  *
```

`$` is equivalent to `[`  (available on Alphanumeric QR Alphabet and on the URI alphabet)
`*` is equivalent to `]`  (available on Alphanumeric QR Alphabet and on the URI alphabet)

Need to make sure fields that use `$` and `*` percent-encode those chars. 

Since the number of fields is fixed within a template, there's no need for a comma character to register the start and end of the next record within the array: it either finishes with a `*` or starts a new record.

**Comment by dpjanes** [comment]
How about encoding the repeats in the same colon-component? That avoids tested components:

`:Jane$John:Doe$Doe:MDL$MDL:23EFG3D5CC$23EFG3D5CC:Passport$Passport:YNW32ND$2SC6223:`

**Comment by vitorpamplona** [comment]
Looks like a lot of work to unpack and figure out which level of the hierarchy each field sits and which parent object it references to

**Comment by vitorpamplona** [comment]
Here's another idea. 

Instead of using $, * to inform nesting, the URL can have the array length in it, as it's own field. 

The JSON above would result in:

```
JXT::data:1:2/Jane/Doe/2/MDL/23EFG3D5CC/Passport/YNW32ND/John/Doe/2MDL/94568DDS1/Passport/2SC6223
```

Breaking down the URI:

```
JXT::data:1:
  2/
    Jane/
    Doe/
    2/
      MDL/23EFG3D5CC/
      Passport/YNW32ND/
    John/
    Doe/
      2MDL/94568DDS1/
      Passport/2SC6223
```

The advantage here is that it's super easy to code.

**Comment by vitorpamplona** [comment]
This was solved by #7
